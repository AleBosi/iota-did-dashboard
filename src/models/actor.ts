// /models/actor.ts

export interface Actor {
  id: string; // DID unico (es: "did:iota:evm:0x...")
  name: string;
  role: "admin" | "azienda" | "creator" | "operatore" | "macchinario";
  aziendaId?: string;      // DID dell’azienda di appartenenza
  seed?: string;           // non esportare mai in VC!
  publicKey?: string;      // per interoperabilità/validazione
  createdAt?: string;      // opzionale, ISO date
  updatedAt?: string;      // opzionale

  // AGGIUNGI QUESTO CAMPO:
  vcIds?: string[];        // Lista degli ID delle VC associate a questo attore
}
