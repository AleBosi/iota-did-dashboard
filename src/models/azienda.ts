// /models/azienda.ts
export interface Azienda {
  id: string;         // DID (es: did:iota:evm:...)
  name: string;
  seed: string;       // per login e firma
  legalInfo?: {
    vat?: string;         // partita IVA o VAT
    lei?: string;         // Legal Entity Identifier (EBSI ready)
    address?: string;
    email?: string;
    country?: string;
    [key: string]: any;   // altri campi liberi
  };
  creators: string[];     // array di DID
  operatori: string[];
  macchinari: string[];
  createdAt?: string;     // ISO
  updatedAt?: string;     // ISO
}
