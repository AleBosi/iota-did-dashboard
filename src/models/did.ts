// /models/did.ts
export interface DidDocument {
  id: string; // es: "did:iota:evm:0x123..."
  controller?: string[];
  verificationMethod?: {
    id: string;
    type: string;
    publicKeyMultibase: string;
  }[];
  authentication?: string[];
  service?: {
    id: string;
    type: string;
    serviceEndpoint: string;
  }[];
  // Altri campi secondo specifica, se serve
}
