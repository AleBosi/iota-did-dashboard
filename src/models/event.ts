// /models/event.ts

export interface Event {
  id: string;
  productId: string;         // ID prodotto collegato (ora obbligatorio)
  operatoreId: string;       // ID operatore coinvolto
  macchinarioId: string;     // ID macchinario coinvolto
  type: string;              // es: "Assemblaggio", "Controllo", "Assegnazione"
  description: string;
  date: string;              // ISO
  creatorId: string;         // DID del creator che ha creato lâ€™evento
  done?: boolean;            // true se eseguito
  bomComponent?: string;     // opzionale, componente specifico
  proofId?: string;          // se firmerai evento come VC
  vcIds?: string[];          // VC associate/notarizzate collegate a questo evento
}
