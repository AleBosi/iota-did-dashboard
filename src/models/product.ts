// /models/product.ts
export interface Product {
  id: string;              // ID univoco locale/app
  did: string;             // DID prodotto (es: did:iota:evm:0x...)
  name: string;            // Nome commerciale o etichetta prodotto
  typeId?: string;         // riferimento a ProductType.id
  serial?: string;         // seriale prodotto
  owner?: string;          // DID del proprietario corrente
  bom?: { id: string; name: string }[]; // lista componenti (Bill of Materials)
  credentials?: string[];  // lista di VC associate (ID delle VC)
  description?: string;    // opzionale
}
