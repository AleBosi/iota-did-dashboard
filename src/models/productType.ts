// /models/productType.ts
export interface ProductType {
  id: string;               // es: GTIN, oppure UUID univoco
  name: string;
  description?: string;
}
