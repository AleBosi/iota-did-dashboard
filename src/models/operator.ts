// /models/operator.ts
import { Actor } from "./actor";

export interface Operator extends Actor {
  // aggiungi campi solo se servono SOLO agli operatori
}
