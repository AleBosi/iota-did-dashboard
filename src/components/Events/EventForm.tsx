import { useState } from "react";
import { Product } from "../../models/product";
import { Actor } from "../../models/actor";
import { Event } from "../../models/event";
import { generateDid } from "../../utils/didUtils";
import { issueVC } from "../../utils/vcHelpers";
import { saveItem } from "../../utils/storageHelpers";
import AssignmentManager from "../Actors/Creator/AssignmentManager";

interface Props {
  prodotti: Product[];
  actors: Actor[];
  eventi: Event[]; // tutti gli eventi già creati (per assignment)
  creatorDid: string;
  onCreate?: (event: any) => void;
}

const EventForm: React.FC<Props> = ({
  prodotti,
  actors,
  eventi,
  creatorDid,
  onCreate,
}) => {
  const [productId, setProductId] = useState("");
  const [operatoreId, setOperatoreId] = useState("");
  const [macchinarioId, setMacchinarioId] = useState("");
  const [type, setType] = useState("");
  const [description, setDescription] = useState("");
  const [bomComponent, setBomComponent] = useState("");
  const [done, setDone] = useState(false);
  const [vcPreview, setVcPreview] = useState<any | null>(null);
  const [assignments, setAssignments] = useState<any[]>([]);

  // Gestione assegnazione tramite AssignmentManager
  const handleAssign = (ass: { eventId: string; operatoreId?: string; macchinarioId?: string }) => {
    setAssignments(prev => [...prev, ass]);
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (!productId || !operatoreId || !macchinarioId || !type) return;

    const prodotto = prodotti.find(p => p.productId === productId);
    const operatore = actors.find(a => a.id === operatoreId && a.role === "operatore");
    const macchinario = actors.find(a => a.id === macchinarioId && a.role === "macchinario");

    if (!prodotto || !operatore || !macchinario) return;

    const eventCore = {
      eventId: generateDid(),
      product: {
        productId: prodotto.productId,
        did: prodotto.did,
        serial: prodotto.serial,
        typeId: prodotto.typeId,
      },
      operatore: {
        did: operatore.did,
        name: operatore.name,
      },
      macchinario: {
        did: macchinario.did,
        name: macchinario.name,
      },
      creator: creatorDid,
      type,
      description,
      date: new Date().toISOString(),
      done,
      bomComponent: bomComponent || undefined,
    };

    // Emissione VC centrata sull'evento
    const vc = issueVC(
      ["VerifiableCredential", "ProductEventCredential"],
      creatorDid,
      eventCore
    );

    const event = {
      ...eventCore,
      vc,
      vcId: vc.id,
      credentials: [vc],
    };

    saveItem(`Event:${event.eventId}`, event);
    saveItem(`VC:${vc.id}`, vc);
    setVcPreview(vc);
    onCreate?.(event);

    setProductId("");
    setOperatoreId("");
    setMacchinarioId("");
    setType("");
    setDescription("");
    setBomComponent("");
    setDone(false);
  };

  return (
    <form onSubmit={handleSubmit} className="p-4 bg-white rounded-2xl shadow flex flex-col gap-2">
      <label>
        Prodotto:
        <select value={productId} onChange={e => setProductId(e.target.value)} required>
          <option value="">-- Seleziona prodotto --</option>
          {prodotti.map(p => (
            <option key={p.productId} value={p.productId}>
              {p.serial} ({p.typeId})
            </option>
          ))}
        </select>
      </label>
      <AssignmentManager
        eventi={eventi}
        actors={actors}
        onAssign={handleAssign}
        assignments={assignments}
      />
      {/* Permetti anche la selezione manuale (per fallback/test) */}
      <label>
        Operatore (diretto):
        <select value={operatoreId} onChange={e => setOperatoreId(e.target.value)}>
          <option value="">-- Seleziona operatore --</option>
          {actors.filter(a => a.role === "operatore").map(op => (
            <option key={op.id} value={op.id}>
              {op.name} ({op.id})
            </option>
          ))}
        </select>
      </label>
      <label>
        Macchinario (diretto):
        <select value={macchinarioId} onChange={e => setMacchinarioId(e.target.value)}>
          <option value="">-- Seleziona macchinario --</option>
          {actors.filter(a => a.role === "macchinario").map(mac => (
            <option key={mac.id} value={mac.id}>
              {mac.name} ({mac.id})
            </option>
          ))}
        </select>
      </label>
      <label>
        Tipo evento:
        <input
          placeholder="Tipo evento"
          value={type}
          onChange={e => setType(e.target.value)}
          required
          className="border px-2 py-1 rounded"
        />
      </label>
      <label>
        Descrizione:
        <input
          placeholder="Dettagli"
          value={description}
          onChange={e => setDescription(e.target.value)}
          className="border px-2 py-1 rounded"
        />
      </label>
      <label>
        Componente BOM:
        <input
          placeholder="(opzionale)"
          value={bomComponent}
          onChange={e => setBomComponent(e.target.value)}
          className="border px-2 py-1 rounded"
        />
      </label>
      <label>
        Eseguito:
        <input
          type="checkbox"
          checked={done}
          onChange={e => setDone(e.target.checked)}
        />
      </label>
      <button type="submit" className="bg-green-600 text-white px-4 py-1 rounded">
        Aggiungi evento
      </button>
      {vcPreview && (
        <div className="mt-4 p-2 bg-gray-50 rounded-xl border text-xs">
          <div className="font-bold mb-1">✅ Evento VC emessa:</div>
          <pre className="overflow-x-auto">{JSON.stringify(vcPreview, null, 2)}</pre>
        </div>
      )}
    </form>
  );
};

export default EventForm;
