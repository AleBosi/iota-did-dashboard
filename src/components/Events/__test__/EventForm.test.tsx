import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import EventForm from "../EventForm";
import { Actor } from "../../../models/actor";
import { Product } from "../../../models/product";
import { Event } from "../../../models/event";

const mockProduct: Product = {
  productId: "PRD-001",
  typeId: "TYPE-A",
  did: "did:example:prod111",
  serial: "XYZ123",
  owner: "did:example:company001",
  children: [],
};

const mockOperatore: Actor = {
  id: "user222",
  did: "did:example:user222",
  name: "Mario Rossi",
  role: "operatore",
};

const mockMacchinario: Actor = {
  id: "machine333",
  did: "did:example:machine333",
  name: "Linea 1",
  role: "macchinario",
};

const mockEventi: Event[] = [
  {
    id: "EV-01",
    type: "avvio",
    description: "Avvio linea",
    productId: "PRD-001",
    operatoreId: "user222",
    macchinarioId: "machine333",
    creatorId: "did:example:creator999",
    date: new Date().toISOString(),
    done: false,
  },
];

describe("EventForm VC-centric", () => {
  it("emette evento e VC, richiama onCreate e resetta i campi", async () => {
    const handleCreate = jest.fn();
    render(
      <EventForm
        prodotti={[mockProduct]}
        actors={[mockOperatore, mockMacchinario]}
        eventi={mockEventi}
        creatorDid="did:example:creator999"
        onCreate={handleCreate}
      />
    );

    // Seleziona prodotto
    fireEvent.change(screen.getByLabelText(/Prodotto/i), {
      target: { value: mockProduct.productId },
    });
    // Seleziona operatore (select manuale)
    fireEvent.change(screen.getByLabelText(/Operatore \(diretto\)/i), {
      target: { value: mockOperatore.id },
    });
    // Seleziona macchinario (select manuale)
    fireEvent.change(screen.getByLabelText(/Macchinario \(diretto\)/i), {
      target: { value: mockMacchinario.id },
    });
    // Tipo evento
    fireEvent.change(screen.getByPlaceholderText(/Tipo evento/i), {
      target: { value: "Montaggio" },
    });
    // Descrizione
    fireEvent.change(screen.getByPlaceholderText(/Dettagli/i), {
      target: { value: "Assemblaggio iniziale" },
    });
    // Componente BOM
    fireEvent.change(screen.getByPlaceholderText(/\(opzionale\)/i), {
      target: { value: "Comp-01" },
    });
    // Eseguito (checkbox)
    fireEvent.click(screen.getByLabelText(/Eseguito/i));

    // Submit
    fireEvent.click(screen.getByRole("button", { name: /Aggiungi evento/i }));

    // Attendi chiamata callback e badge VC preview
    await waitFor(() => {
      expect(handleCreate).toHaveBeenCalled();
      expect(screen.getByText(/VC emessa/i)).toBeInTheDocument();
    });

    // Verifica struttura evento creato
    const event = handleCreate.mock.calls[0][0];
    expect(event.product.productId).toBe(mockProduct.productId);
    expect(event.operatore.name).toBe(mockOperatore.name);
    expect(event.macchinario.name).toBe(mockMacchinario.name);
    expect(event.type).toBe("Montaggio");
    expect(event.done).toBe(true);
    expect(event.vc).toBeDefined();
    expect(event.vc["@context"]).toBeDefined();

    // Verifica reset campi (il select prodotto torna vuoto)
    expect(screen.getByPlaceholderText(/Tipo evento/i)).toHaveValue("");
    expect(screen.getByPlaceholderText(/Dettagli/i)).toHaveValue("");
    expect(screen.getByPlaceholderText(/\(opzionale\)/i)).toHaveValue("");
  });

  it("matcha lo snapshot del form", () => {
    const { asFragment } = render(
      <EventForm
        prodotti={[mockProduct]}
        actors={[mockOperatore, mockMacchinario]}
        eventi={mockEventi}
        creatorDid="did:example:creator999"
        onCreate={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
