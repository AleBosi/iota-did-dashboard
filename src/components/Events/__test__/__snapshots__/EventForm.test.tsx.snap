// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`EventForm VC-centric matcha lo snapshot del form 1`] = `
<DocumentFragment>
  <form
    class="p-4 bg-white rounded-2xl shadow flex flex-col gap-2"
  >
    <label>
      Prodotto:
      <select
        required=""
      >
        <option
          value=""
        >
          -- Seleziona prodotto --
        </option>
        <option
          value="PRD-001"
        >
          XYZ123 (TYPE-A)
        </option>
      </select>
    </label>
    <div
      class="p-4 border rounded bg-gray-50 mb-4"
    >
      <h3
        class="font-semibold mb-2"
      >
        Assegna Operatore e Macchinario a un Evento
      </h3>
      <div
        class="flex flex-col gap-3 mb-4"
      >
        <select
          class="border rounded px-2 py-1"
        >
          <option
            value=""
          >
            Seleziona Evento
          </option>
          <option
            value="EV-01"
          >
            avvio - Avvio linea
          </option>
        </select>
        <select
          class="border rounded px-2 py-1"
        >
          <option
            value=""
          >
            Seleziona Operatore
          </option>
          <option
            value="user222"
          >
            Mario Rossi (did:example:user222)
          </option>
        </select>
        <select
          class="border rounded px-2 py-1"
        >
          <option
            value=""
          >
            Seleziona Macchinario
          </option>
          <option
            value="machine333"
          >
            Linea 1 (did:example:machine333)
          </option>
        </select>
        <button
          class="bg-blue-600 text-white px-4 py-1 rounded mt-2"
          disabled=""
        >
          Assegna
        </button>
      </div>
    </div>
    <label>
      Operatore (diretto):
      <select>
        <option
          value=""
        >
          -- Seleziona operatore --
        </option>
        <option
          value="user222"
        >
          Mario Rossi (user222)
        </option>
      </select>
    </label>
    <label>
      Macchinario (diretto):
      <select>
        <option
          value=""
        >
          -- Seleziona macchinario --
        </option>
        <option
          value="machine333"
        >
          Linea 1 (machine333)
        </option>
      </select>
    </label>
    <label>
      Tipo evento:
      <input
        class="border px-2 py-1 rounded"
        placeholder="Tipo evento"
        required=""
        value=""
      />
    </label>
    <label>
      Descrizione:
      <input
        class="border px-2 py-1 rounded"
        placeholder="Dettagli"
        value=""
      />
    </label>
    <label>
      Componente BOM:
      <input
        class="border px-2 py-1 rounded"
        placeholder="(opzionale)"
        value=""
      />
    </label>
    <label>
      Eseguito:
      <input
        type="checkbox"
      />
    </label>
    <button
      class="bg-green-600 text-white px-4 py-1 rounded"
      type="submit"
    >
      Aggiungi evento
    </button>
  </form>
</DocumentFragment>
`;
